static/js/message_list_view.js

```diff
@@ -1445,4 +1445,87 @@ export class MessageListView {
             message_container.status_message = false;
         }
     }
+
+    update_sticky_recipient_headers() {
+        const rows_length = this._rows.size;
+        if (!rows_length) {
+            /* No headers are present */
+            return;
+        }
+        /* Intentionally remove sticky headers class here to make calculations simpler. */
+        $(".sticky_header").removeClass("sticky_header");
+        /* visible_top is navbar top position + height for us. */
+        const visible_top = $message_viewport.message_viewport_info().visible_top;
+        /* We need date to be properly visible on the header, so partially visible headers
+           who are about to be scrolled out of view are not acceptable. */
+        const header_about_to_be_partially_hidden = visible_top - 1;
+
+        function is_sticky(header) {
+            const header_props = header.getBoundingClientRect();
+            const margin_between_recipient_rows = 10;
+            const header_is_sticky_or_about_to_be_sticky =
+                visible_top + header_props.height + margin_between_recipient_rows;
+            /* If two headers next to each other are completely visible
+               (message header at top has no visible content), we don't mind showing
+               date on any of them. Which header is chosen will depend on which
+               comes first when iterating on the headers. */
+            return (
+                header_about_to_be_partially_hidden <= header_props.top &&
+                header_props.top <= header_is_sticky_or_about_to_be_sticky
+            );
+        }
+
+        const $table = rows.get_table(this.table_name);
+        const $headers = $table.find(".message_header");
+        const iterable_headers = $headers.toArray();
+        /* To get to the sticky header ASAP, we take help from how far has user scrolled. */
+        if ($message_viewport.scrollTop() > $message_viewport.height() / 2) {
+            iterable_headers.reverse();
+        }
+        let sticky_header;
+        for (const header of iterable_headers) {
+            if (is_sticky(header)) {
+                sticky_header = header;
+                break;
+            }
+        }
```

Comment:
This should use a binary search (`_.sortedIndexBy`).

Or, better, can we find the message row *first* with `document.elementsFromPoint`, then find the sticky header from the message?

Reply:
`document.elementsFromPoint` is very risky to use since there are a lot of UI state we would have to take into account.

I don't understand how we can use `_.sortedIndexBy` or binary search here. I fear we would end up doing more work in common cases since the user is mostly looking at the latest unread message, in which case, the sticky header is just is 0-5 message headers away.

Reply:
“Risky” in what way? You’re already using `document.elementsFromPoint` below, and you’re already doing a complicated computation with UI state in `is_sticky` above. If that’s risky, I don’t see this being any worse.

If there are 400 messages visible, binary search does 8 or 9 tests, while your linear search does up to 200. The potential gain if linear search happens to do 0–5 tests is small, while the potential loss is large.

Reply:
I tried implementing it using `document.elementsFromPoint`, seems to be working fine, it would be great if you can check too!

Reply:
We should definitely profile this bit of code while doing scrolling around (ideally including things like using the Mac trackpad acceleration in "all messages"), renarrowing, etc., once we've settled on how we think it's cleanest to do it.  

I haven't read this carefully, but one risk with whatever we're doing here is that removing the `sticky` class at the start of this function and then trying to read what is in a given location might trigger an intermediate repaint or something that could be slow?

Reply:
> I haven't read this carefully, but one risk with whatever we're doing here is that removing the sticky class at the start of this function and then trying to read what is in a given location might trigger an intermediate repaint or something that could be slow?

Since we are not changing the position of any element by removing the `sticky_header` class, it most likely will repaint the date part of the message row if it is still visible in the narrow which is harmless. I will do a profile to confirm though.
```
        &.sticky_header {
            .recipient_row_date {
                display: block;
            }
        }
```

Reply:
I removed the method which used `document.elementsFromPoint` and went back to my original method and applied binary search to it as recommended by @andersk. This made some of the comments below not relveant anymore.